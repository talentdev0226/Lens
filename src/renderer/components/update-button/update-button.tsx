/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import styles from "./styles.module.scss";

import type { HTMLAttributes } from "react";
import React, { useState } from "react";
import { Menu, MenuItem } from "../menu";
import { cssNames } from "../../utils";
import type { IconProps } from "../icon";
import { Icon } from "../icon";

interface UpdateButtonProps extends HTMLAttributes<HTMLButtonElement> {
  warningLevel?: "light" | "medium" | "high";
  update: () => void;
}

export function UpdateButton({ warningLevel, update, id }: UpdateButtonProps) {
  const buttonId = id ?? "update-lens-button";
  const menuIconProps: IconProps = { material: "update", small: true };
  const [opened, setOpened] = useState(false);

  const toggle = () => {
    setOpened(!opened);
  };

  if (!warningLevel) {
    return null;
  }

  return (
    <>
      <button
        data-testid="update-button"
        data-warning-level={warningLevel}
        id={buttonId}
        className={cssNames(styles.updateButton, {
          [styles.warningHigh]: warningLevel === "high",
          [styles.warningMedium]: warningLevel === "medium",
        })}
      >
        Update
        <Icon material="arrow_drop_down" className={styles.icon}/>
      </button>
      <Menu
        usePortal
        htmlFor={buttonId}
        isOpen={opened}
        close={toggle}
        open={toggle}
      >
        <MenuItem
          icon={menuIconProps}
          onClick={update}
          data-testid="update-lens-menu-item"
        >
          Relaunch to Update Lens
        </MenuItem>
      </Menu>
    </>
  );
}
