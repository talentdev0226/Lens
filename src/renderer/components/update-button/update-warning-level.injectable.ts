/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import { computed } from "mobx";
import { now as reactiveDateNow } from "mobx-utils";
import updateDownloadedDateTimeInjectable from "../../../common/application-update/update-downloaded-date-time/update-downloaded-date-time.injectable";

const updateWarningLevelInjectable = getInjectable({
  id: "update-warning-level",

  instantiate: (di) => {
    const updateDownloadedDateTime = di.inject(updateDownloadedDateTimeInjectable);

    return computed(() => {
      const downloadedAt = updateDownloadedDateTime.value.get();

      if (!downloadedAt) {
        return "";
      }

      const ONE_DAY = 1000 * 60 * 60 * 24;

      const downloadedAtTimestamp = new Date(downloadedAt).getTime();
      const currentDateTimeTimestamp = reactiveDateNow(ONE_DAY);

      const elapsedTime = currentDateTimeTimestamp - downloadedAtTimestamp;

      const elapsedDays = elapsedTime / ONE_DAY;

      if (elapsedDays < 20) {
        return "light";
      }

      if (elapsedDays >= 20 && elapsedDays < 25) {
        return "medium";
      }

      return "high";
    });
  },
});

export default updateWarningLevelInjectable;
