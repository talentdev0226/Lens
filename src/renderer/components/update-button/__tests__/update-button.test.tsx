/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { render, act } from "@testing-library/react";
import React from "react";
import { UpdateButton } from "../update-button";
import "@testing-library/jest-dom/extend-expect";

const update = jest.fn();

describe("<UpdateButton/>", () => {
  beforeEach(() => {
    update.mockClear();
  });

  it("should not render if no warning level prop passed", () => {
    const { queryByTestId } = render(<UpdateButton update={update} />);

    expect(queryByTestId("update-button")).not.toBeInTheDocument();
  });

  it("should render if warning level prop passed", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    expect(getByTestId("update-button")).toMatchSnapshot();
  });

  it("should open menu when clicked", async () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    const button = getByTestId("update-button");

    act(() => button.click());

    expect(getByTestId("update-lens-menu-item")).toBeInTheDocument();
  });

  it("should call update function when menu item clicked", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    const button = getByTestId("update-button");

    act(() => button.click());

    const menuItem = getByTestId("update-lens-menu-item");

    menuItem.click();

    expect(update).toHaveBeenCalled();
  });

  it("should have light warning level", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="light" />);

    const button = getByTestId("update-button");

    expect(button.dataset.warningLevel).toBe("light");
  });

  it("should have medium warning level", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="medium" />);

    const button = getByTestId("update-button");

    expect(button.dataset.warningLevel).toBe("medium");
  });

  it("should have high warning level", () => {
    const { getByTestId } = render(<UpdateButton update={update} warningLevel="high" />);

    const button = getByTestId("update-button");

    expect(button.dataset.warningLevel).toBe("high");
  });
});
