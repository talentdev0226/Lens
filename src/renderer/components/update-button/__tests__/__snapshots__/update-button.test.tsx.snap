// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateButton/> should render if warning level prop passed 1`] = `
<button
  class="updateButton"
  data-testid="update-button"
  data-warning-level="light"
  id="update-lens-button"
>
  Update
  <i
    class="Icon icon material focusable"
  >
    <span
      class="icon"
      data-icon-name="arrow_drop_down"
    >
      arrow_drop_down
    </span>
  </i>
</button>
`;
